// av/src/sources/browser/http.js
// Generated by CoffeeScript 1.10.0
(function() {
  var AVBuffer, EventEmitter, HTTPSource,
    extend = function(child, parent) { for (var key in parent) { if (hasProp.call(parent, key)) child[key] = parent[key]; } function ctor() { this.constructor = child; } ctor.prototype = parent.prototype; child.prototype = new ctor(); child.__super__ = parent.prototype; return child; },
    hasProp = {}.hasOwnProperty;

  EventEmitter = require('../../core/events');

  AVBuffer = require('../../core/buffer');

  HTTPSource = (function(superClass) {
    extend(HTTPSource, superClass);

    function HTTPSource(url, opts) {
      this.url = url;
      this.opts = opts != null ? opts : {};
      this.chunkSize = 1 << 20;
      this.inflight = false;
      if (this.opts.length) {
        this.length = this.opts.length;
      }
      this.reset();
    }

    HTTPSource.prototype.start = function() {
      if (this.length) {
        if (!this.inflight) {
          return this.loop();
        }
      }
      this.inflight = true;
      this.xhr = new XMLHttpRequest();
      this.xhr.onload = (function(_this) {
        return function(event) {
          _this.length = parseInt(_this.xhr.getResponseHeader("Content-Length"));
          _this.inflight = false;
          return _this.loop();
        };
      })(this);
      this.xhr.onerror = (function(_this) {
        return function(err) {
          _this.pause();
          return _this.emit('error', err);
        };
      })(this);
      this.xhr.onabort = (function(_this) {
        return function(event) {
          return _this.inflight = false;
        };
      })(this);
      this.xhr.open("HEAD", this.url, true);
      return this.xhr.send(null);
    };

    HTTPSource.prototype.loop = function() {
      var endPos;
      if (this.inflight || !this.length) {
        return this.emit('error', 'Something is wrong in HTTPSource.loop');
      }
      this.inflight = true;
      this.xhr = new XMLHttpRequest();
      this.xhr.onload = (function(_this) {
        return function(event) {
          var buf, buffer, i, j, ref, txt;
          if (_this.xhr.response) {
            buf = new Uint8Array(_this.xhr.response);
          } else {
            txt = _this.xhr.responseText;
            buf = new Uint8Array(txt.length);
            for (i = j = 0, ref = txt.length; 0 <= ref ? j < ref : j > ref; i = 0 <= ref ? ++j : --j) {
              buf[i] = txt.charCodeAt(i) & 0xff;
            }
          }
          buffer = new AVBuffer(buf);
          _this.offset += buffer.length;
          _this.emit('data', buffer);
          if (_this.offset >= _this.length) {
            _this.emit('end');
          }
          _this.inflight = false;
          if (!(_this.offset >= _this.length)) {
            return _this.loop();
          }
        };
      })(this);
      this.xhr.onprogress = (function(_this) {
        return function(event) {
          return _this.emit('progress', (_this.offset + event.loaded) / _this.length * 100);
        };
      })(this);
      this.xhr.onerror = (function(_this) {
        return function(err) {
          _this.emit('error', err);
          return _this.pause();
        };
      })(this);
      this.xhr.onabort = (function(_this) {
        return function(event) {
          return _this.inflight = false;
        };
      })(this);
      this.xhr.open("GET", this.url, true);
      this.xhr.responseType = "arraybuffer";
      endPos = Math.min(this.offset + this.chunkSize, this.length - 1);
      this.xhr.setRequestHeader("If-None-Match", "webkit-no-cache");
      this.xhr.setRequestHeader("Range", "bytes=" + this.offset + "-" + endPos);
      this.xhr.overrideMimeType('text/plain; charset=x-user-defined');
      return this.xhr.send(null);
    };

    HTTPSource.prototype.pause = function() {
      var ref;
      this.inflight = false;
      return (ref = this.xhr) != null ? ref.abort() : void 0;
    };

    HTTPSource.prototype.reset = function() {
      this.pause();
      return this.offset = 0;
    };

    return HTTPSource;

  })(EventEmitter);

  module.exports = HTTPSource;

}).call(this);
